File: Python3.x/371-Sum_of_Two_Integers.py
# Runtime: 36 ms, faster than 83.47% of Python3 online submissions for Sum of Two Integers.
# Memory Usage: 13.3 MB, less than 8.57% of Python3 online submissions for Sum of Two Integers.
class Solution:
    def getSum(self, a: int, b: int) -> int:
        MAX, MIN, mask = 0x7FFFFFFF, 0x80000000, 0xFFFFFFFF
        while b != 0:
            a, b = (a ^ b) & mask, ((a & b) << 1) & mask
        return a if a <= MAX else ~(a ^ mask)  # ～是按位取反

File: Algorithms/C++/371-Sum_of_Two_Integers.cpp
# 对数运算的性质
// Runtime: 0 ms, faster than 100.00% of C++ online submissions for Sum of Two Integers.
// Memory Usage: 8.4 MB, less than 5.15% of C++ online submissions for Sum of Two Integers.
class Solution {
public:
    int getSum(int a, int b) 
    {
        if(a < b) swap(a, b);
        if(a==INT_MAX && b==INT_MIN) return -1;
        return (log(pow(2, b) * pow(2, a)) / log(2));
    }
};
